File: src/main/js/formulas/math/ACOS.ts
import { NumException } from "../../errors/NumException";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class ACOS extends AbstractFormula {
  static SELF: ACOS = new ACOS();
  NAME = FormulaName.ACOS;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLength(values.length, 1, this.NAME);
    const n = Converters.toNumber(Converters.first(this.collateralLookup(origin, values[0])));
    if (n < -1 || n > 1) {
      return new NumException(
        `Function ACOS parameter 1 value is ${n}. Valid values are between -1 and 1 inclusive.`
      );
    }
    return Math.acos(n);
  }
}

File: src/main/java/io/protobase/f7/formulas/math/ACOS.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;

import java.util.function.BiFunction;
import java.util.function.Function;

public class ACOS extends AbstractFormula {
  public static FormulaName NAME = FormulaName.ACOS;
  public static ACOS SELF = new ACOS();

  public ACOS() {
    super();
  }

  public ACOS(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkLength(values.length, 1, NAME);
    return Math.acos(Converters.toDouble(Converters.first(collateralLookup.apply(origin, values[0]))));
  }
}
